exporter/exporterhelper/persistent_queue.go

```diff
@@ -0,0 +1,470 @@
+	} else {
+		wcs.writeIndex = writeIndexIf.(itemIndex)
+	}
+}
+
+// loop is the main loop that handles fetching items from the persistent buffer
+func (pcs *persistentContiguousStorage) loop() {
+	// We want to run it here so it's not blocking
+	pcs.loadLeftProcessedItems(context.Background())
+
+	for {
+		for {
+			req, found := pcs.getNextItem(context.Background())
+			if found {
+				select {
+				case <-pcs.stopChan:
+					return
+				case pcs.reqChan <- req:
+				}
+			} else {
+				select {
+				case <-pcs.stopChan:
+					return
+				case <-time.After(pcs.retryDelay):
+				}
+			}
+		}
+	}
+}
+
+// get returns the request channel that all the requests will be send on
+func (pcs *persistentContiguousStorage) get() <-chan request {
+	return pcs.reqChan
+}
+
+// size returns the number of currently available items, which were not picked by consumers yet
+func (pcs *persistentContiguousStorage) size() int {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+	return int(pcs.writeIndex - pcs.readIndex)
+}
+
+func (pcs *persistentContiguousStorage) stop() {
+	pcs.logger.Debug("stopping persistentContiguousStorage", zap.String(zapQueueNameKey, pcs.queueName))
+	pcs.stopOnce.Do(func() {
+		close(pcs.stopChan)
+	})
+}
+
+// Put marshals the request and puts it into the persistent queue
+func (pcs *persistentContiguousStorage) put(req request) error {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	ctx := context.Background()
+
+	itemKey := pcs.buildItemKey(pcs.writeIndex)
+	if !pcs._clientSet(ctx, itemKey, req, requestToBytes) {
+		return errStoringItemToQueue
+	}
+
+	pcs.writeIndex++
+	if !pcs._clientSet(ctx, writeIndexKey, pcs.writeIndex, itemIndexToBytes) {
+		return errUpdatingIndex
+	}
+
+	return nil
+}
+
+// getNextItem pulls the next available item from the persistent storage; if none is found, returns (nil, false)
+func (pcs *persistentContiguousStorage) getNextItem(ctx context.Context) (request, bool) {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	if pcs.readIndex < pcs.writeIndex {
+		index := pcs.readIndex
+		itemKey := pcs.buildItemKey(index)
+		// Increase here, so despite errors it would still progress
+		pcs.readIndex++
+
+		pcs.proceedToNextItem(ctx)
+		pcs.markProcessedItem(ctx, index)
+
+		req := pcs._clientGet(ctx, itemKey, pcs.bytesToRequest).(request)
+		if req == nil {
+			return nil, false
+		}
+
+		req.setOnProcessingFinished(func() {
+			pcs.removeProcessedItem(ctx, index)
+		})
+		return req, true
+	}
+
+	return nil, false
+}
+
+// loadLeftProcessedItems gets the unprocessed items from the persistent queues and moves them back to the queue
+func (pcs *persistentContiguousStorage) loadLeftProcessedItems(ctx context.Context) {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	pcs.logger.Debug("checking if there are items left by consumers")
+	processedItemsIf := pcs._clientGet(ctx, currentlyProcessedItemsKey, bytesToItemIndexArray)
+	if processedItemsIf == nil {
+		return
+	}
+	processedItems := processedItemsIf.([]itemIndex)
```

Comment:
```suggestion
	processedItems ,ok := processedItemsIf.([]itemIndex)
```

...

otherwise if it's not a `[]itemIndex` then it blows up?